packit/local_project.py

```diff
@@ -452,12 +476,51 @@ def checkout_pr(self, pr_id: Union[str, int]):
             "merge-requests" if is_gitlab else "pull", pr_id
         )
         remote_name = self.remote or "origin"
-        local_ref = f"refs/remotes/{remote_name}/pr/{pr_id}"
-        local_branch = f"pr/{pr_id}"
-        self.git_repo.remotes[remote_name].fetch(f"{remote_ref}:{local_ref}")
-        self.git_repo.create_head(local_branch, f"{remote_name}/{local_branch}")
-        self.git_repo.branches[local_branch].checkout()
-        logger.info(f"Checked out commit {self.git_repo.head.commit}")
+        local_ref = f"refs/remotes/{remote_name}/pr-changes/{pr_id}"
+        local_branch = f"pr-changes/{pr_id}"
+
+        self._fetch_as_branch(remote_ref, local_ref, local_branch)
+
+        head_commit = self.git_repo.branches[local_branch].commit
+        logger.info(
+            f"Checked out commit\n"
+            f"({shorten_commit_hash(head_commit.hexsha)})\t{head_commit.summary}"
+        )
+
+    def merge_pr(self, pr_id: Union[str, int]) -> None:
+        """
+        Merge given PR into target branch. Fetches and switches to base branch
+        (where changes from the PR are to be merged) and then merges branch with
+        changes from the PR.
+
+        Args:
+            pr_id: ID of the PR we are merging.
+
+        Raises:
+            PackitException: In case merge fails.
+        """
+        remote = self.remote or "origin"
+        pr = self.git_project.get_pr(int(pr_id))
+
+        local_target_branch = f"pr/{pr_id}"
+        self._fetch_as_branch(
+            f"+refs/heads/{pr.target_branch}",
+            f"refs/remotes/{remote}/pr/{pr_id}",
+            local_target_branch,
+        )
+        self.git_repo.branches[local_target_branch].checkout()
+        target_branch = self.git_repo.branches[local_target_branch]
+
+        commit_sha = shorten_commit_hash(target_branch.commit.hexsha)
+        logger.info(
+            f"Merging ({pr.target_branch}) with commit:\n"
+            f"({commit_sha})\t{target_branch.commit.summary}"
+        )
+        try:
+            self.git_repo.git.merge(f"pr-changes/{pr_id}")
+        except GitCommandError as ex:
+            logger.error(f"Merge failed with: {ex}")
+            raise PackitException(ex)
```

Comment:
we totally need to figure out what to do here before merging

I'd personally incline to `reset --hard $PR` and mention this in the check description - it should also be clear in github's interface that the PR cannot be merged

Reply:
GitHub/GitLab yells by itself that you have merge conflicts, iirc it probably checks the same way `git merge` works, it's also mentioned in description

Reply:
packit/packit-service#1206, not sure how much we can rely on it (with regards to _test merge commit_)

Reply:
we can certainly try

I'd probably raise PackitCannotRebaseException here, catch it in p-s and set the neutral state there.